bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java

```diff
@@ -0,0 +1,395 @@
+        if (authenticate()) {
+            updateStatus(ThingStatus.ONLINE);
+            poll();
+        }
+    }
+
+    protected abstract void poll();
+
+    protected abstract void handleCommandAuthenticated(final ChannelUID channelUID, final Command command);
+
+    protected void updateStatus(ThingStatus status) {
+        updateStatus(status, ThingStatusDetail.NONE);
+    }
+
+    protected void updateStatus(ThingStatus status, ThingStatusDetail detail) {
+        updateStatus(status, detail, null);
+    }
+
+    protected abstract void updateStatus(ThingStatus status, ThingStatusDetail detail, final String message);;
+
+    private static class Authentication {
+        public Authentication(final String privateKey, final String cookie) {
+            this.privateKey = privateKey;
+            this.cookie = cookie;
+        }
+
+        public String privateKey;
+        public String cookie;
+    }
+
+    public void start(ScheduledExecutorService scheduler) {
+        try {
+            pollFuture = scheduler.scheduleWithFixedDelay(this::poller, 0, DETECT_POLL_S, TimeUnit.SECONDS);
+        } catch (Exception e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.HANDLER_INITIALIZING_ERROR,
+                    "Unexpected internal error.");
+            logger.error("Failed to start http client or scheduler.");
+        }
+    }
+
+    public void stop() {
+        if (pollFuture != null) {
+            pollFuture.cancel(true);
+        }
+    }
+
+    public void reset() {
+        // force re-authentication before the next action.
+        authentication = null;
+
+        // the configuration may be correct.
+        // if it was incorrect, the `authenticate()`-method will set the flag.
+        invalidConfiguration = false;
+    }
+
+    public void handleCommand(final ChannelUID channelUID, final Command command) {
+        if (authenticate()) {
+            handleCommandAuthenticated(channelUID, command);
+        }
+    }
+
+    public void setPin(final String pin) {
+        this.pin = pin;
+    }
+
+    public void setIp(final String ip) {
+        this.url = "http://" + ip + "/HNAP1/";
+    }
+
+    private SOAPElement moduleParameters(final String moduleID) throws SOAPException {
+        return soapFactory.createElement("ModuleID").addTextNode(moduleID);
+    }
+
+    public String queryStringState(final String action, final String name, final String module) {
+        try {
+            final Document document = post(action, newMessage(action, moduleParameters(module)), authentication);
+            return getElementByTagName(document, name);
+        } catch (CommunicationException e) {
+            logger.error("Failed to get quantity (communication error).");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+        } catch (SOAPException e) {
+            logger.error("Unexpected internal error.");
+            updateStatus(ThingStatus.OFFLINE);
+        }
```

Comment:
You could include the exceptions message as third parameter to `updateStatus()` to display the error in the UI.
You could remove the logging statement, as `updateStatus()` already does the logging.